# Dynamic Programming

Today we **Solve once, reuse forever.**

## What is Dynamic Programming?

### Informal definition

Dynamic programming solves complex problems by breaking them into smaller pieces, solving each piece once, and reusing those results instead of repeating work.

### Academically sounder definition

Dynamic programming is a method for solving problems by decomposing them into overlapping subproblems, solving each subproblem once, and systematically storing and reusing their results to construct an optimal global solution.

## Why should you care?

Besides being more practical (read applicable more often), Dynamic Programming problems are favorite of various job interviews. It is almost guaranteed that you would encounter a Dynamic Programming problem when interviewing at the larger companies.


## Richard E. Bellman - father of Dynamic Programming

![Bellman](https://upload.wikimedia.org/wikipedia/en/7/7a/Richard_Ernest_Bellman.jpg)

https://en.wikipedia.org/wiki/Richard_E._Bellman

**Richard E. Bellman (1920–1984)** — American mathematician and one of the founders of dynamic programming.
* **Dynamic Programming (1950s):** Formalized a method for solving optimization problems by breaking them into **overlapping subproblems** guided by the **principle of optimality**.
* **Core Contribution:** Provided a general framework for **multistage decision processes**, enabling efficient algorithm design via memoization and tabulation.
* **Algorithmic Impact:** His ideas underpin key algorithms such as the **Bellman–Ford shortest path algorithm**, which handles graphs with negative edge weights.
* **Broader Fields:** Made significant contributions to **control theory**, **operations research**, **stochastic processes**, **numerical analysis**, and **differential equations**.
* **Interdisciplinary Influence:** Bridged theoretical mathematics with practical computation, influencing economics, engineering, and computer science.
* **Modern Relevance:** Foundational to **reinforcement learning** and **optimal control**, where decision-making over time is central.


### Dynamic Programming != Regular Programming

The term dynamic programming (or simply DP) can be a bit confusing to newcomers. Both of the words are
used in a different way than most might expect. Programming here refers to making a set of choices (as in “linear
programming”) and thus has more in common with the way the term is used in, say, television, than in writing
computer programs. Dynamic simply means that things change over time—in this case, that each choice depends
on the previous one. In other words, this “dynamicism” has little to do with the program you’ll write and is just a
description of the problem class. In Bellman’s own words, “I thought dynamic programming was a good name. It was
something not even a Congressman could object to. So I used it as an umbrella for my activities

* The core technique of DP -> caching
* Decompose your problem recursively/inductively (usual)
* allow overlap between the subproblems.
* Plain recursive solution xponential number of times -> caching trims away waste
* result is usually both an impressively efficient algorithm and a greater insight into the problem.


Commonly, DP algorithms turn the recursive formulation upside down, making it iterative and filling out some
data structure (such as a multidimensional array) step by step.

* Another option well suited to high-level languages such as Python—is to implement the recursive formulation directly but to cache the return
values.
* If a call is made more than once with the same arguments, the result is simply returned directly from the
cache. This is known as **memoization**

## Case 1 - Fibonacci numbers

The Fibonacci problem asks us to compute the *n*-th number in the Fibonacci sequence — a sequence of integers defined by a simple recurrence relation:

* (F(0) = 0)
* (F(1) = 1)
* (F(n) = F(n-1) + F(n-2)) for (n >= 2)

In other words, each number is the sum of the two preceding ones.

**Task:**
Given a non-negative integer `n`, compute the value of (F(n)).

**Example:**

* (F(2) = 1)
* (F(3) = 2)
* (F(4) = 3)
* (F(5) = 5)

This problem serves as a foundational example for studying dynamic programming because:

* It has **overlapping subproblems** (the same values are recomputed many times)
* It exhibits **optimal substructure** (the solution depends on solutions to smaller instances)

We will explore multiple approaches to solving this problem, starting from a naive recursive solution and progressing toward more efficient dynamic programming techniques.


### Naive Recursive Solution


In [12]:
## Fibonacci
def fib(i): # finding i-th member in our Fibonacci chain 0,1,1,2,3,5,8,13,21,34,55,89 - we skip 0 here, think why?
    if i < 2: # base cases - first two numbers are 0,1
        return i # so return 0,1
    else:
        return fib(i-1) + fib(i-2) # we have two recursive calls, but our problem space is not going down by half

In [13]:
fib(0),fib(1),fib(2),fib(3),fib(4),fib(5)

(0, 1, 1, 2, 3, 5)

In [15]:
# sidenote: Fibonaccy ratios approach golden mean
for n in range(2,20):
    print(fib(n), fib(n-1), round(fib(n)/fib(n-1),8))

1 1 1.0
2 1 2.0
3 2 1.5
5 3 1.66666667
8 5 1.6
13 8 1.625
21 13 1.61538462
34 21 1.61904762
55 34 1.61764706
89 55 1.61818182
144 89 1.61797753
233 144 1.61805556
377 233 1.61802575
610 377 1.61803714
987 610 1.61803279
1597 987 1.61803445
2584 1597 1.61803381
4181 2584 1.61803406


![Fib Spiral](https://upload.wikimedia.org/wikipedia/commons/thumb/9/93/Fibonacci_spiral_34.svg/500px-Fibonacci_spiral_34.svg.png)

https://en.wikipedia.org/wiki/Golden_ratio

In [16]:
# so far so good?
fib(30)

832040

In [17]:
fib(35)

9227465

In [18]:
fib(36)

14930352

### Fibonacci Recursion

https://stackoverflow.com/questions/35959100/explanation-on-fibonacci-recursion
![FibTree](https://i.stack.imgur.com/QVSdv.png)

### Memoization solution

Some modern programming languages offer a prebuilt memoization solution. This means you do not have to build your own cache from scratch.

Python is among them using functools.lru_cache in standard library.

### lru_cache wrapper

Newer versions of Python have a decorator that can perform memoization. It’s called lru_cache
There is even a maxsize parameter that allows you to limit the size of the cache. This is useful if you know that
you’ll only need the last n results. The name stands for “least recently used,” and it means that if the cache is full, the less used values are discarded first.

In [20]:
import functools
# https://stackoverflow.com/questions/1988804/what-is-memoization-and-how-can-i-use-it-in-python
@functools.lru_cache(maxsize=None) #by default only 128 latest so by setting it to None I cache everything - this is a decorator
def fib_memo(num):
    if num < 2:
        return num
    else:
        return fib_memo(num-1) + fib_memo(num-2)
# in 3.9 Python there is a newer @functools.cache - similar idea

In [21]:
fib(35)

9227465

In [22]:
fib_memo(35)

9227465

In [ ]:
fib_memo(35)

In [ ]:
%%timeit
fib(35)

86.3 ns ± 27.8 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


In [23]:
fib_memo(100)

354224848179261915075

### SideQuest - Explore how limiting cache size will affect performance

In [27]:
# let's limit cache to only say 20 values
# https://stackoverflow.com/questions/1988804/what-is-memoization-and-how-can-i-use-it-in-python
@functools.lru_cache(maxsize=20) #by default only 128 latest so by setting it to None I cache everything - this is a decorator
def fib_lim20(num):
    if num < 2:
        return num
    else:
        return fib_lim20(num-1) + fib_lim20(num-2)
# in 3.9 Python there is a newer @functools.cache - similar idea

In [28]:
fib_lim20(40)

102334155

In [ ]:
## TODO investigate on how cache is cleared on cache_lru

In [ ]:
# so the problem with TOP-DOWN memoization is that we are still left with recursive calls going over our stack limit

### Solving the stack overflow problem?

If you run the memoized version for larger number you will still run into stack overflow problem that is common in situations where recursive stack is limited.

You could raise the stack size, or you could use another approach - tabulation.

In [ ]:
# we could try the build up solution - meaning BOTTOM-UP method
# this will usually be an iterative(looping) solution so no worries about stack


### Iterative version of Fibonacci with full table

In [30]:
# silly iterative version
# useful if you want to return ALL fibs
def fib_it(n): # so n will be 0 based
    fibs = [0,1] #so we are going to build our answers
    # lets pretend we do not know of any formulas and optimizations
#     n += 1 # fix this
    if n < 2:
        return fibs[n] # off by one errors
    ndx = 2
    while ndx <= n:
        fibs.append(fibs[ndx-1]+fibs[ndx-2]) # so I am building a 1-d table(array/list) of answers
        # so appending to list incurs some cost as well
        # in fib we could have of course just stored 2 previous values instead of all
        ndx+=1
    return fibs[n] # again off by one indexing 0 based in python and 1 based in our function


In [31]:
fib_it(2)

1

In [32]:
fib_it(4)

3

In [33]:
fib_it(100)

354224848179261915075

In [34]:
%%timeit
fib_memo(100)

69.9 ns ± 0.944 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


In [35]:
%%timeit
fib_it(100)

10.8 µs ± 61.3 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


In [ ]:
# Why is iterative solution slower?

# Think of a couple of reasons here

3.16 µs ± 49.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


In [ ]:
len(str(fib_it(10_000)))

2090

### Improving Tabular Solution - Store Less!

In [ ]:
# so one way we could improve is that we do not need to store all this knowledge about previous solutions,
# (unless we were building a table of ALL solutions)

In [37]:
# so we only need to store 2 values, but this would be improving constant and solving our space complexity
# instead of needing O(n) to store our fib(n), we just need O(1) - since we are onl storing two values
def fib_v2(n):
    prev, cur = 0, 1 # so we save memory by not storing all previous values
    if n <= 1:
        return prev
    ndx = 2
    while ndx <= n:
        prev, cur = cur, prev+cur # python makes it easy to assign 2 values at once with tuple unpacking, right side eval first
        # if your language does not support it you can simply use a third temp variable
        ndx += 1
    return cur

In [38]:
fib_v2(5),fib_v2(6)

(5, 8)

In [40]:
%%timeit
fib_v2(100)

6.89 µs ± 2.25 µs per loop (mean ± std. dev. of 7 runs, 100000 loops each)


## Case 2 - Pascal's triangle
![Triangle](https://upload.wikimedia.org/wikipedia/commons/0/0d/PascalTriangleAnimated2.gif)

The combinatorial
meaning of C(n,k) is the number of k-sized subsets you can get from a set of size n.

In mathematics, a combination is a selection of items from a collection, such that the order of selection does not matter (unlike permutations). https://en.wikipedia.org/wiki/Combination

In [ ]:
# this is horrible again we have 2 recursive calls for each call
def C(n,k):
    # first base case
    if k == 0: return 1 # means we have chosen all k items
    if n == 0: return 0 # means we have no items to choose from
    return C(n-1,k-1) + C(n-1,k)

In [ ]:
C(2,0),C(2,1),C(2,2)

(1, 2, 1)

In [ ]:
C(3,0),C(3,1),C(3,2),C(3,3)

(1, 3, 3, 1)

In [ ]:
C(4,0),C(4,1),C(4,2),C(4,3),C(4,4)

(1, 4, 6, 4, 1)

In [ ]:
for k in range(6):
    print(C(5,k), end=" ")
print()

1 5 10 10 5 1 


In [ ]:
%%timeit
C(20,12)

170 ms ± 24.7 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [ ]:
import functools

In [41]:
# so exact same function just with caching of results
@functools.lru_cache(maxsize=None)
def C_mem(n,k):
    if k == 0: return 1
    if n == 0: return 0
    return C_mem(n-1,k-1) + C_mem(n-1,k)

In [42]:
C_mem(20,12)

125970

In [43]:
%%timeit
C_mem(20,12)

133 ns ± 42.6 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


In [ ]:
%%timeit
C_mem(25,15)

65.4 ns ± 12.9 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


In [44]:
# how many different ways we can place 8 identical pieces (queens) on chess board
C_mem(64,8)

4426165368

In [ ]:
%%timeit
C_mem(50,30)

131 ns ± 5.35 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)


In [ ]:
C_mem(64,8) # so BRUTE FORCE for 8-queens problem would require checking this many

4426165368

In [ ]:
2**32 # no relation to above number

4294967296

In [ ]:
%%timeit
C_mem(20,12)

203 ns ± 30 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


So 3 Million times faster
And it will only get worse with larger n and k values


In [ ]:
C(22,10)  # keeps getting slower

646646

In [ ]:
C_mem(30,22)

5852925

In [ ]:
C_mem(500,127)

46437038934385738875060685074580974915827322393712987687159475139635977046776280972687848152763132582010929859049061968000

In [ ]:
C_mem(2000,1555) # so we calculated how many different ways we can pick 1555 items out of set of 2000 items

537425311263286649045728031247848043926269386315713439394349281522386287881044189299660257402655939762820833577618210552868619936076461996950343651583167622419765426925933775006289767911413157852444651898746557509579506062674421789656825030627507059629199364370311140152613280678386654811275586813136941829060081126272639830187540208190808703691937737524007173610897100607592784614616627680955021048820164545102485799315069769979712239992887424317738387672000

In [ ]:
C_mem(49,6)

13983816

In [ ]:
C_mem(55,7)  # so our odds in lotteries are not so great...

202927725

You may at times want to rewrite your code to make it iterative. This
can make it faster, and you avoid exhausting the stack if the recursion depth gets excessive. There’s another reason, too:
The iterative versions are often based on a specially constructed cache, rather than the generic “dict keyed by parameter
tuples” used in my @memo.

This means that you can sometimes use more efficient structures, such as the multidimensional
arrays of NumPy (this is a C type array dense and efficient), or even just nested lists.

This custom cache design
makes it possible to do use DP in more low-level languages(ahem C, C++), where general, abstract solutions such as our @memo decorator
are often not feasible.

Note that even though these two techniques often go hand in hand, you are certainly free to use an
iterative solution with a more generic cache or a recursive one with a tailored structure for your subproblem solutions.

Let’s reverse our algorithm, filling out Pascal’s triangle directly.

In [ ]:
from collections import defaultdict  # we could use a regular dict , just less work with defaultdict
# in your language check what collections are available, C#, Java have very nice collections libraries
# careful with C++ collections, check complexity on some of them, it might not be O(1) - map for one (which turns out to be ordered)
# you'd need to look up documentation for that particular structure - so in C++ you might need to use unordered map

In [ ]:
def pascal_up(n,k):
    # so tabulated solution, we are going to build a table of solutions
    Cit = defaultdict(int)
    # Cit = {} # turns out going to plain dictionary did not help at all, slow down by 10%
    # careful from off by one errors
    for row in range(n+1):
        Cit[row,0] = 1 # we could have used previous row to calculate this one, technically just need first one
        for col in range(1,k+1): # looking like O(n*k) space and time complexity here right?
            # I am using get to avoid looking at edge cases (where we have no value to look up)
            Cit[row,col] = Cit.get((row-1,col-1),0) + Cit.get((row-1,col),0)
    return Cit[n,k]

### Pascal triangle with iterative DP - tabulation

In [ ]:
pascal_up(20,12)

125970

In [ ]:
%%timeit
pascal_up(20,12)

125 µs ± 25.7 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


In [ ]:
pascal_up(50,30)

47129212243960

In [ ]:
%%timeit
pascal_up(50,30)

681 µs ± 5.81 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [ ]:
%%timeit
C_mem(20,12)

131 ns ± 6.31 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)


In [ ]:
pascal_up(200,120)

1647278652451762678788128833110870712983038446517480945400

In [ ]:
C_mem(200,120)

1647278652451762678788128833110870712983038446517480945400

In [ ]:
pascal_up(2024,1223)

1978313580909322814404279742888220639083711979061915952790417145826485941327234894939258385780824686228079384862407027961433025056766438463773323839416940536781231172166724228912771210164695158929585213737757529910987728315522103983421673060914798337366517226742260910135847960153174204828837881201328813992496688218062280491920325152745752467152549279836765394857076973819686255193997158217396454096616616197813730448592629973705108771904401411312180534910702836373199696360608671550707240571399665331895768867458771375296689485866658969198724600312085207901774291234960477899482724866200

In [ ]:
# C_mem(2000,1255) # so we see the need for iterative version - other way would be to allow tail call optimization in functional languages

In [ ]:
# so we could futher save memory for our pascal_up by only saving the needed information meaning we only need the previous row

### Pascal triangle with iterative DP - tabulation

In [ ]:
pascal_up(20,12)

125970

In [ ]:
%%timeit
pascal_up(20,12)

137 µs ± 4.18 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [ ]:
pascal_up(50,30)

47129212243960

In [ ]:
%%timeit
pascal_up(50,30)

406 µs ± 5.44 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


In [ ]:
%%timeit
C_mem(20,12)

131 ns ± 6.31 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)


In [ ]:
pascal_up(200,120)

1647278652451762678788128833110870712983038446517480945400

In [ ]:
C_mem(200,120)

1647278652451762678788128833110870712983038446517480945400

In [ ]:
pascal_up(2000,1255)

59109593628908414948382028796066195825322294401884421608567282370732677108447726193711784730302760990299603214712874545403948907216441186822204871444990120418511945136706510520055755556735802821658571978418751352273298267569828345746978809441973750241286775122448947529540564006233186974990594534845090378195822798749749716850238790136589693799060708520531078121496569151588079913232831392320592160010660575061801016625866394043795717433667107028907325906172066327055484845091441072359350112683890957329354388664535777738465246096416216139097702868890896959249389851651200

In [ ]:
# C_mem(2000,1255) # so we see the need for iterative version - other way would be to allow tail call optimization in functional languages

In [ ]:
# so we could futher save memory for our pascal_up by only saving the needed information meaning we only need the previous row

## Difference between TOP-DOWN (with memoization) and BOTTOM-UP (with filling up DP table)

Basically the same thing is going on. The main difference is that we need to figure out which cells in the cache
need to be filled out, and we need to find a safe order to do it in so that when we’re about to calculate C[row,col], the
cells C[row-1,col-1] and C[row-1,col] are already calculated. With the memoized function, we needn’t worry about
either issue: It will calculate whatever it needs recursively.

In [ ]:
## Back to LIS

In [ ]:
# so recursive memoized solution - TOP/DOWN
def rec_lis(seq): # Longest increasing subseq.
    @functools.lru_cache(maxsize=None)
    def L(cur): # Longest ending at seq[cur]
        res = 1 # Length is at least 1 - because well because one number is always going to be a solution to non empty list
        # so for example reversely sorted list would have a solution 1 - because the numbers are going down
        for pre in range(cur): # Potential predecessors
            if seq[pre] <= seq[cur]: # A valid (smaller) predec.
                res = max(res, 1 + L(pre)) # Can we improve the solution?
        return res
    return max(L(i) for i in range(len(seq))) # The longest of them all

In [ ]:
rec_lis([3,1,0,2,4])

3

In [ ]:
naive_list([3,1,0,2,4,7,-6,9])

(1, 2, 4, 7, 9)

In [ ]:
rec_lis([3,1,0,2,4,7,-6,9])  # so 5 should be answer here

5

In [ ]:
# so recursive memoized solution - TOP/DOWN
# TODO add sequence passing
def rec_lis_full(seq): # Longest increasing subseq.
    @functools.lru_cache(maxsize=None)
    def L(cur): # Longest ending at seq[cur]
        res = 1 # Length is at least 1
        for pre in range(cur): # Potential predecessors
            if seq[pre] <= seq[cur]: # A valid (smaller) predec.
                res = max(res, 1 + L(pre)) # Can we improve the solution?
        return res
    return max(L(i) for i in range(len(seq))) # The longest of them all

In [ ]:
# tabulated solution
# tabulation is the classical dynamic programming approach - BOTTOM - UP
def basic_lis(seq, debug=True):
    L = [1] * len(seq)
    for cur, val in enumerate(seq):
        for pre in range(cur): # so two nested loops smells quadratic right?
            if seq[pre] < val:
                L[cur] = max(L[cur], 1 + L[pre])
                if debug:
                    print(L)
    return max(L) # linear lookup

In [ ]:
# TODO add iterative sequence passing

In [ ]:
basic_lis([3,1,2,0,4])

[1, 1, 2, 1, 1]
[1, 1, 2, 1, 2]
[1, 1, 2, 1, 2]
[1, 1, 2, 1, 3]
[1, 1, 2, 1, 3]


3

In [ ]:
basic_lis([3,1,0,2,4])

[1, 1, 1, 2, 1]
[1, 1, 1, 2, 1]
[1, 1, 1, 2, 2]
[1, 1, 1, 2, 2]
[1, 1, 1, 2, 2]
[1, 1, 1, 2, 3]


3

In [ ]:
basic_lis([3,1,0,2,4,7,-6,9])

5

In [ ]:
for i,n in enumerate("Valdis",start=100):  # just a Pythonic way of adding index to sequences
    print(i,n)

100 V
101 a
102 l
103 d
104 i
105 s


In [ ]:
basic_lis([3,1,0,2,4,7,9,6])

[1, 1, 1, 2, 1, 1, 1, 1]
[1, 1, 1, 2, 1, 1, 1, 1]
[1, 1, 1, 2, 2, 1, 1, 1]
[1, 1, 1, 2, 2, 1, 1, 1]
[1, 1, 1, 2, 2, 1, 1, 1]
[1, 1, 1, 2, 3, 1, 1, 1]
[1, 1, 1, 2, 3, 2, 1, 1]
[1, 1, 1, 2, 3, 2, 1, 1]
[1, 1, 1, 2, 3, 2, 1, 1]
[1, 1, 1, 2, 3, 3, 1, 1]
[1, 1, 1, 2, 3, 4, 1, 1]
[1, 1, 1, 2, 3, 4, 2, 1]
[1, 1, 1, 2, 3, 4, 2, 1]
[1, 1, 1, 2, 3, 4, 2, 1]
[1, 1, 1, 2, 3, 4, 3, 1]
[1, 1, 1, 2, 3, 4, 4, 1]
[1, 1, 1, 2, 3, 4, 5, 1]
[1, 1, 1, 2, 3, 4, 5, 2]
[1, 1, 1, 2, 3, 4, 5, 2]
[1, 1, 1, 2, 3, 4, 5, 2]
[1, 1, 1, 2, 3, 4, 5, 3]
[1, 1, 1, 2, 3, 4, 5, 4]


5

In [ ]:
# so basic_lis is tabulating and using two nested loops - thus O(n^2) (again not all nested loops will be n^2 but most are)
# can we do better for this increasing subsequence length

A crucial insight is that if more than one predecessor terminate subsequences of length m, it doesn’t matter which
one of them we use—they’ll all give us an optimal answer. Say, we want to keep only one of them around; which one
should we keep? The only safe choice would be to keep the smallest of them, because that wouldn’t wrongly preclude
any later elements from building on it. So let’s say, inductively, that at a certain point we have a sequence end of
endpoints, where end[idx] is the smallest among the endpoints we’ve seen for increasing subsequences of length idx+1
(we’re indexing from 0). Because we’re iterating over the sequence, these will all have occurred earlier than our current
value, val. All we need now is an inductive step for extending end, finding out how to add val to it. If we can do that, at
the end of the algorithm len(end) will give us the final answer—the length of the longest increasing subsequence.

This devilishly clever little algorithm was first was first described by Michael L. Fredman in 1975

## Optimal Longest Increasing Subsequence
https://www.sciencedirect.com/science/article/pii/0012365X7590103X

In [ ]:
# https://docs.python.org/3/library/bisect.html - basically we use binary search to find insertion points
from bisect import bisect  # provides support for maintaining a list in sorted order without having to sort the list after each insertion.
def lis(seq): # Longest increasing subseq.
    end = [] # End-values for all lengths
    for val in seq: # Try every value, in order
        idx = bisect(end, val) # no way to make this constant but this is logn
        if idx == len(end): # Can we build on an end val?
            end.append(val) # Longest seq. extended
        else:
            end[idx] = val # Prev. endpoint reduced
    return len(end) # The longest we found

In [ ]:
lis([3,1,0,2,4,7,9,6])

5

## So Top-Down
* Create a recursive solution
* memoize it - (cache it!)

## Bottom-Up
* figure out a way of storing previous results
* tabulate(calculate) up
* you might not need to store the full table maybe just previous row - can save space

In [ ]:
## TODO tabulated solution to LIS problem that returns the actual sequence

## Coin change solution using dynamic bottom-up programming

Idea is to use a table of previous results to calculate the next one
Next solution has to be a sum of two previous solutions - so we can use the previous results to calculate the next one

Let's remember how our greedy solution failed on some cases

* For example Greedy would fail on [1, 3, 4] and 6 - where correct solution is 3,3 but greedy would give 4,1,1
* Another fail would be on [1,10,25] and 31 - greedy would give 25,1,1,1,1,1,1 and correct solution is 10,10,10,1

Instead we will use a dynamic programming solution - we will use a table to store the results of the previous solutions and use them to calculate the next one

In [ ]:
# we will use a list to store our results
# we will have a sorted list of coins for each amount
# if we have a coin matching the amount we will add it to the list
def find_minimum_coins(amount, coins=(1,5,10,25)):
    table = [0] * (amount + 1) # so we are going to store our results here
    for i in range(1, amount + 1): # so we are going to go through all amounts
        table[i] = min(table[i - c] + 1 for c in coins if c <= i) # so we are going to find the minimum number of coins for each amount
    return table[amount], table

# let's test it on 31
coin_cnt, full_solution = find_minimum_coins(31)
# print full solution with amounts
print(*list(enumerate(full_solution)), sep="\n")

(0, 0)
(1, 1)
(2, 2)
(3, 3)
(4, 4)
(5, 1)
(6, 2)
(7, 3)
(8, 4)
(9, 5)
(10, 1)
(11, 2)
(12, 3)
(13, 4)
(14, 5)
(15, 2)
(16, 3)
(17, 4)
(18, 5)
(19, 6)
(20, 2)
(21, 3)
(22, 4)
(23, 5)
(24, 6)
(25, 1)
(26, 2)
(27, 3)
(28, 4)
(29, 5)
(30, 2)
(31, 3)


In [ ]:
# lets try it for amount 15, and coins being (1,7,10)
coin_cnt, full_solution = find_minimum_coins(15, (1,7,10))
# print full solution with amounts
print(*list(enumerate(full_solution)), sep="\n")

(0, 0)
(1, 1)
(2, 2)
(3, 3)
(4, 4)
(5, 5)
(6, 6)
(7, 1)
(8, 2)
(9, 3)
(10, 1)
(11, 2)
(12, 3)
(13, 4)
(14, 2)
(15, 3)


In [ ]:
# now let's test 31 but use only coins (1,10,25)
coin_cnt, full_solution = find_minimum_coins(31, (1,10,25))
# print full solution with amounts
print(*list(enumerate(full_solution)), sep="\n")

(0, 0)
(1, 1)
(2, 2)
(3, 3)
(4, 4)
(5, 5)
(6, 6)
(7, 7)
(8, 8)
(9, 9)
(10, 1)
(11, 2)
(12, 3)
(13, 4)
(14, 5)
(15, 6)
(16, 7)
(17, 8)
(18, 9)
(19, 10)
(20, 2)
(21, 3)
(22, 4)
(23, 5)
(24, 6)
(25, 1)
(26, 2)
(27, 3)
(28, 4)
(29, 5)
(30, 3)
(31, 4)


In [ ]:
# lets get max int from sys
import sys
sys.maxsize, 2**63-1 # which is 2^63 - 1 because we also have 0 in our range


(9223372036854775807, 9223372036854775807)

In [ ]:
# let's modify the function to store the actual coins used
def find_minimum_coins(amount, coins=(1,5,10,25)):
    table = [sys.maxsize] * (amount + 1) # so we are going to store our results here
    coins_used = [[]] * (amount + 1) # so we are going to store our results here
    for i in range(1, amount + 1): # so we are going to go through all amounts
        # table[i], coins_used[i] = min((table[i - c] + 1, c) for c in coins if c <= i) # so we are going to find the minimum number of coins for each amount
        if i in coins: # exact match so only 1 coin needed
            table[i] = 1
            coins_used[i] = [i]
        else: # now we need to find a minimum by combining previous solutions
            for n in range(i//2+1): # you could just go throuh i but this is a bit faster
                # we check previous solutions for minimum
                if table[n] + table[i-n] < table[i]:
                    table[i] = table[n] + table[i-n]
                    coins_used[i] = coins_used[n] + coins_used[i-n]
    return table[amount], coins_used

# test it on 31
coin_cnt, coins_used = find_minimum_coins(31)
# print coin_cnt
print(coin_cnt)
# print coins used
print(coins_used)

3
[[], [1], [1, 1], [1, 1, 1], [1, 1, 1, 1], [5], [1, 5], [1, 1, 5], [1, 1, 1, 5], [1, 1, 1, 1, 5], [10], [1, 10], [1, 1, 10], [1, 1, 1, 10], [1, 1, 1, 1, 10], [5, 10], [1, 5, 10], [1, 1, 5, 10], [1, 1, 1, 5, 10], [1, 1, 1, 1, 5, 10], [10, 10], [1, 10, 10], [1, 1, 10, 10], [1, 1, 1, 10, 10], [1, 1, 1, 1, 10, 10], [25], [1, 25], [1, 1, 25], [1, 1, 1, 25], [1, 1, 1, 1, 25], [5, 25], [1, 5, 25]]


In [ ]:
# now let's check with 1,10,25 on 31
coin_cnt, coins_used = find_minimum_coins(31, (1,10,25))
# print coin_cnt
print(coin_cnt)
# print coins used
print(*coins_used, sep="\n")

4
[]
[1]
[1, 1]
[1, 1, 1]
[1, 1, 1, 1]
[1, 1, 1, 1, 1]
[1, 1, 1, 1, 1, 1]
[1, 1, 1, 1, 1, 1, 1]
[1, 1, 1, 1, 1, 1, 1, 1]
[1, 1, 1, 1, 1, 1, 1, 1, 1]
[10]
[1, 10]
[1, 1, 10]
[1, 1, 1, 10]
[1, 1, 1, 1, 10]
[1, 1, 1, 1, 1, 10]
[1, 1, 1, 1, 1, 1, 10]
[1, 1, 1, 1, 1, 1, 1, 10]
[1, 1, 1, 1, 1, 1, 1, 1, 10]
[1, 1, 1, 1, 1, 1, 1, 1, 1, 10]
[10, 10]
[1, 10, 10]
[1, 1, 10, 10]
[1, 1, 1, 10, 10]
[1, 1, 1, 1, 10, 10]
[25]
[1, 25]
[1, 1, 25]
[1, 1, 1, 25]
[1, 1, 1, 1, 25]
[10, 10, 10]
[1, 10, 10, 10]


## Exercises on Memoization

### Stair Climbing (Counting Problem)

**Core idea:**
Count the number of distinct ways to reach step `n` using a fixed set of allowed step sizes (typically `{1, 2}` or `{1, 2, 3}`).

#### Problem formulation

* Input: integer `n` (target step)
* Allowed moves: e.g., `{1, 2}`
* Output: number of distinct sequences of moves reaching exactly `n`

#### Recursive structure

* Let `f(n)` be the number of ways to reach step `n`
* Recurrence:

  * `f(n) = f(n-1) + f(n-2)` (for steps `{1,2}`)
* Base cases:

  * `f(0) = 1` (one way: do nothing)
  * `f(n < 0) = 0`

#### Memoization insight

* The same subproblems (`f(n-1)`, `f(n-2)`, etc.) are recomputed many times
* Cache results for each `n` to reduce exponential recursion to linear time

#### Extensions (for exploration)

* Allow `{1, 2, 3}` steps → `f(n) = f(n-1) + f(n-2) + f(n-3)`
* Introduce forbidden steps (e.g., broken stairs)
* Return one valid path instead of just the count

---

### Rod Cutting (Optimization Problem)

**Core idea:**
Maximize profit by cutting a rod into pieces, given a price table for each possible length.

#### Problem formulation

* Input:

  * Array `price[i]` = value of rod of length `i`
  * Integer `n` = total rod length
* Output: maximum obtainable value

#### Recursive structure

* Let `r(n)` be the maximum revenue for length `n`
* Recurrence:

  * `r(n) = max(price[i] + r(n - i))` for all `1 ≤ i ≤ n`
* Base case:

  * `r(0) = 0`

#### Memoization insight

* Many recursive calls recompute `r(k)` for the same `k`
* Cache computed values of `r(k)` to avoid redundant work
* Reduces exponential recursion to quadratic time

#### Extensions (for exploration)

* Reconstruct actual cuts (not just value)
* Add cost per cut (penalize excessive splitting)
* Compare with greedy strategies and show failure cases

---

### Edit Distance (String Transformation)

**Core idea:**
Compute the minimum number of operations required to transform one string into another.

#### Problem formulation

* Input: two strings `A` and `B`
* Allowed operations:

  * Insert
  * Delete
  * Substitute
* Output: minimum number of operations

#### Recursive structure

* Let `d(i, j)` be the edit distance between:

  * first `i` characters of `A`
  * first `j` characters of `B`
* Recurrence:

  * If characters match:
    `d(i, j) = d(i-1, j-1)`
  * Otherwise:
    `d(i, j) = 1 + min(`

    * `d(i-1, j)`   (delete)
    * `d(i, j-1)`   (insert)
    * `d(i-1, j-1)` (substitute)
      `)`
* Base cases:

  * `d(0, j) = j`
  * `d(i, 0) = i`

#### Memoization insight

* Subproblems `(i, j)` repeat frequently
* Cache results in a dictionary or 2D structure
* Converts exponential recursion into polynomial time `O(n·m)`

#### Extensions (for exploration)

* Reconstruct the sequence of edits
* Assign different costs to operations
* Compare with Longest Common Subsequence (LCS)

---

## Related topics (brief pointers)

* Bottom-up tabulation vs top-down memoization
* State definition and dimensionality in DP
* Reconstruction of optimal solutions
* Time–space tradeoffs in DP implementations


## Algorithms that use Dynamic Programming

### Graph algorithms

* Shortest path algorithms
* Longest path algorithms
* Minimum spanning tree algorithms
* Maximum flow algorithms

### String algorithms

* Longest common subsequence
* Longest increasing subsequence
* Edit distance

### Combinatorial algorithms

* Binomial coefficients - Pascal's triangle
* Fibonacci numbers
* Factorials

### Geometric algorithms

* Convex hull
* Closest pair of points
* Largest empty rectangle

## More Exercises

### Coin Change (Number of Ways)

**Problem Statement**
Given a set of coin denominations and a target amount, compute the **number of distinct ways** to form that amount using **unlimited copies** of each coin.

#### Input

* `coins`: list of positive integers (denominations)
* `target`: non-negative integer

#### Output

* Integer: number of combinations

#### Key Constraints / Assumptions

* Order **does not matter** (i.e., combinations, not permutations)
* Each coin can be used infinitely many times

#### DP Formulation

* **State:**
  `dp[a]` = number of ways to make amount `a`
* **Base Case:**
  `dp[0] = 1` (one way to form zero: choose nothing)
* **Transition:**
  For each coin ( c ):
  [
  dp[a] += dp[a - c] \quad \text{for } a \ge c
  ]

#### Iteration Order (Critical)

```python
for coin in coins:
    for a in range(coin, target + 1):
        dp[a] += dp[a - coin]
```

#### Complexity

* Time: ( O(n \cdot target) )
* Space: ( O(target) )

#### Common Pitfalls

* Reversing loop order → counts permutations instead of combinations
* Forgetting base case `dp[0] = 1`

---

### 0/1 Knapsack

**Problem Statement**
Given `n` items with weights and values, determine the **maximum total value** that can be obtained without exceeding a given capacity. Each item can be taken **at most once**.

#### Input

* `weights[0..n-1]`
* `values[0..n-1]`
* `capacity`

#### Output

* Maximum achievable value

#### DP Formulation

* **State:**
  `dp[i][w]` = maximum value using first `i` items with capacity `w`
* **Base Cases:**

  * `dp[0][w] = 0` (no items)
  * `dp[i][0] = 0` (zero capacity)

#### Transition

For item ( i ):

* If not taken:
  [
  dp[i][w] = dp[i-1][w]
  ]
* If taken (when ( w \ge weight[i] )):
  [
  dp[i][w] = \max(dp[i-1][w],; value[i] + dp[i-1][w - weight[i]])
  ]

#### Table Filling Strategy

* Iterate over items (outer loop)
* Iterate over capacity (inner loop)

#### Space Optimization (1D)

```python
for i in range(n):
    for w in range(capacity, weights[i] - 1, -1):
        dp[w] = max(dp[w], values[i] + dp[w - weights[i]])
```

#### Complexity

* Time: ( O(n \cdot capacity) )
* Space:

  * 2D: ( O(n \cdot capacity) )
  * Optimized: ( O(capacity) )

#### Common Pitfalls

* Iterating capacity **forward** in 1D version → incorrectly allows reuse of items
* Confusing with **unbounded knapsack**

---

### Matrix Chain Multiplication (Advanced)

**Problem Statement**
Given a sequence of matrices, determine the most efficient way to multiply them by choosing optimal parenthesization that minimizes scalar multiplication cost.

#### Input

* Dimensions array `p[0..n]`

  * Matrix ( A_i ) has dimensions ( p[i-1] \times p[i] )

#### Output

* Minimum number of scalar multiplications

#### DP Formulation

* **State:**
  `dp[i][j]` = minimum cost to multiply matrices ( A_i \dots A_j )
* **Base Case:**
  `dp[i][i] = 0` (single matrix requires no multiplication)

#### Transition

For all splits ( k \in [i, j-1] ):
[
dp[i][j] = \min_{k} \left( dp[i][k] + dp[k+1][j] + p[i-1] \cdot p[k] \cdot p[j] \right)
]

#### Table Filling Order

* Increasing chain length ( L = 2 \dots n )
* For each ( (i, j) ) with ( j = i + L - 1 )

#### Iterative Structure

```python
for L in range(2, n+1):
    for i in range(1, n-L+2):
        j = i + L - 1
        dp[i][j] = INF
        for k in range(i, j):
            cost = dp[i][k] + dp[k+1][j] + p[i-1]*p[k]*p[j]
            dp[i][j] = min(dp[i][j], cost)
```

#### Complexity

* Time: ( O(n^3) )
* Space: ( O(n^2) )

#### Conceptual Focus

* **DP over intervals**
* Non-local dependencies (every split must be evaluated)
* Diagonal table filling

#### Common Pitfalls

* Incorrect indexing of dimension array
* Filling table in wrong order (must ensure subproblems solved first)

---

## Related Concepts

* Unbounded vs bounded knapsack
* DP over sequences vs DP over intervals
* Order sensitivity in DP transitions
* Space optimization patterns in tabulation


## References and Additional Readings

## Materials used

* https://jeffe.cs.illinois.edu/teaching/algorithms/book/03-dynprog.pdf

* [Python Algorithms: Mastering Basic Algorithms in the Python Language](https://www.amazon.com/Python-Algorithms-Mastering-Basic-Language/dp/148420056X)

* Chapter on Dynamic Programming - CLRS https://mitpress.mit.edu/9780262046305/introduction-to-algorithms/


## Key References in Dynamic Programming

Note: You do not have to read them all, but reading foundational texts can be surprisingly approachable.

### 1. Foundational Theory

**Bellman, R. (1952).**
*On the theory of dynamic programming.*
*Proceedings of the National Academy of Sciences*, **38**(8), 716–719.
→ Introduces dynamic programming as a general method for solving multistage optimization problems; establishes the core idea of breaking problems into stages.

---

**Bellman, R. (1954).**
*The theory of dynamic programming.*
*Bulletin of the American Mathematical Society*, **60**(6), 503–515.
→ Formalizes the principle of optimality and provides the first rigorous theoretical framework for dynamic programming.

---

**Bellman, R. (1957).**
*Dynamic Programming.* Princeton University Press.
→ The definitive monograph consolidating the theory, including the Bellman equation and numerous applications.

---

### 2. Early Applications and Extensions

**Bellman, R., & Dreyfus, S. E. (1962).**
*Applied Dynamic Programming.* Princeton University Press.
→ Extends DP to practical engineering and operations research problems; introduces computational considerations.

---

**Howard, R. A. (1960).**
*Dynamic Programming and Markov Processes.* MIT Press.
→ Establishes the connection between DP and **Markov Decision Processes (MDPs)**, forming the basis for stochastic control and reinforcement learning.

---

**Blackwell, D. (1962).**
*Discrete dynamic programming.*
*Annals of Mathematical Statistics*, **33**(2), 719–726.
→ Provides rigorous results on optimal policies and convergence in stochastic dynamic programming settings.

---

### 3. Algorithmic Dynamic Programming (Computer Science)

**Held, M., & Karp, R. M. (1962).**
*A dynamic programming approach to the traveling salesman problem.*
*SIAM Journal on Applied Mathematics*, **10**(1), 196–210.
→ Introduces the classic exponential-time DP algorithm for TSP; foundational example of DP over subsets (bitmask DP).

---

**Karp, R. M. (1972).**
*Reducibility among combinatorial problems.*
In R. E. Miller & J. W. Thatcher (Eds.), *Complexity of Computer Computations* (pp. 85–103). Springer.
→ Defines NP-completeness; clarifies when DP is feasible versus when problems are inherently intractable.

---

### 4. Modern Extensions (Learning and Approximation)

**Sutton, R. S. (1988).**
*Learning to predict by the methods of temporal differences.*
*Machine Learning*, **3**(1), 9–44.
→ Connects dynamic programming with **reinforcement learning**, introducing temporal-difference methods as approximate DP.

---

**Bertsekas, D. P., & Tsitsiklis, J. N. (1996).**
*Neuro-Dynamic Programming.* Athena Scientific.
→ Develops approximate dynamic programming using function approximation; foundational for modern deep reinforcement learning.


These works collectively establish:

* **Bellman (1950s):** What DP *is* (principle + recursion)
* **Howard / Blackwell (1960s):** DP under uncertainty
* **Held–Karp (1960s):** DP as an algorithm design technique
* **Karp (1970s):** Limits of DP (complexity theory)
* **Sutton / Bertsekas (modern):** DP as a learning framework


